src/core/blocks/create_block.c

```diff
@@ -0,0 +1,165 @@
+  coinbase_tx->num_outputs = 1;
+  coinbase_tx->outputs = miner_output;
+
+  return coinbase_tx;
+}
+
+/*
+Calculate tX fees for 1 tx
+*/
+unsigned int calc_tx_fees(Transaction *tx){
+  unsigned long total_in = 0;
+  unsigned long total_out = 0;
+  
+  for(unsigned int i = 0; i < tx->num_inputs; i++){
+    UTXO* input_utxo = utxo_pool_find(tx->inputs[i].prev_tx_id, tx->inputs[i].prev_utxo_output);
+    total_in += input_utxo->amt;
+  }
+  for(unsigned int i = 0; i < tx->num_outputs; i++){
+    total_out += tx->outputs->amt;
+  }
+  return total_in - total_out;
+}
+
+
+unsigned int get_txs_from_mempool(Transaction ***tx_pts){
+  unsigned long tx_fees = 0;
+  unsigned int num_included = 0;
+  MemPool *s;
+  unsigned int coinbase = 1;
+
+  unsigned int tx_desired = calc_num_tx_target();
+  unsigned int tx_in_mempool = HASH_COUNT(mempool);
+
+  //Take minimum value of the mempool or the desired size
+  unsigned int tx_to_get = ((tx_desired) < (tx_in_mempool)) ? (tx_desired) : (tx_in_mempool);
+
+  *tx_pts = malloc((tx_to_get+coinbase)*sizeof(Transaction**)); // Shouldn't this be multiplied by 4 because eeach pointer takes up 4 bytes?
+
+  for(s = mempool; s != NULL; s = s->hh.next) {
+    (*tx_pts)[num_included+1] = s->tx; // Leave the first tx for coinbase tx
+    tx_fees += calc_tx_fees(s->tx);
+    num_included++;
+    if(num_included == tx_to_get){
+      break;
+    }
+  }
+
+  // Create the Coinbase TX
+  (*tx_pts)[0] = create_coinbase_tx(tx_fees);
+  
+  unsigned int total_num_txs = num_included+coinbase;
+  // Resize the Transaction memory if the wrong size was created
+  if(num_included < tx_to_get){
+    void *tmp = realloc(*tx_pts, (total_num_txs)*sizeof(Transaction**));
+    if (tmp == NULL) {
+        //realloc didn't work  still points to the og location
+    }
+    else{
+      *tx_pts = tmp;
+    }
+  }
+  return total_num_txs;
+}
+
+unsigned char *hash_all_tx(Transaction **txs, unsigned int num_txs){
+  unsigned char *total_hash = malloc(ALL_TX_HASH_LEN);
+  unsigned char *temp_all = malloc(TX_HASH_LEN*num_txs);
+  
+  for(unsigned int i = 0; i < num_txs; i++){
+    hash_tx(temp_all+i*TX_HASH_LEN, (txs)[i]);
+  }
+
+  hash_sha256(total_hash, temp_all, TX_HASH_LEN*num_txs);
+  free(temp_all);
+  return total_hash;
+}
+
+BlockHeader *create_block_header(Transaction **txs, unsigned int num_txs){
+  BlockHeader *b_header = malloc(sizeof(BlockHeader));
+  unsigned char *prev_header_hash = get_prev_header_hash();
+  unsigned char *all_tx_hash = hash_all_tx(txs, num_txs);
+
+  memcpy(b_header->prev_header_hash, prev_header_hash, ALL_TX_HASH_LEN); 
+  memcpy(b_header->all_tx, all_tx_hash, ALL_TX_HASH_LEN); 
+  b_header->nonce = 0;
+  b_header->timestamp = time(NULL);
+  
+  return b_header;
+}
+
+void create_block(Block *block){
+  Transaction **tx_ptr = malloc(sizeof(Transaction**));
+  block->num_txs = get_txs_from_mempool(&tx_ptr);
+  block->header = *create_block_header(tx_ptr, block->num_txs);
+  block->txs = tx_ptr; 
```

Comment:
I don't follow this at all, could you elaborate. What specifically should be null. Don't understand how there is a leak here?

Reply:
We are currently storing a pointer (`tx_ptr`) that points to some memory (allocated by the malloc call) in `blocks->txs`.

`block->txs` should also be set by the collect txs function so that it points to an array of tx pointers. When this happens, the old pointer is lost and the memory it references leaks.

I think the confusion is with the double pointer. `block->txs` is just a pointer, and already has memory allocated as it is part of the struct. It's no different from outputs (`Output*`) or inputs (`Input*`), where you don't have to allocate memory for the pointer. 

Instead, we allocate memory for what that pointer points to, in this case an array of pointers. This is done in the collect tx function. Here, since we don't have an array to point to, the pointer should just be NULL (like we're making a tx with no inputs, `tx->inputs = NULL`)

When actually allocating the memory, it's the same as any other array. For an array of outputs, we do `malloc(sizeof(Output) * len)`. For an array of tx pointers, we do `malloc(sizeof(Trasaction*) * len)`.

Reply:
```c
  //Transaction **tx_ptr = malloc(sizeof(Transaction**));
  block->txs = NULL;
  block->num_txs = get_txs_from_mempool(block->txs);
  block->header = *create_block_header(tx_ptr, block->num_txs);
```
So something like this instead?